pkg/diagnose/cni.go

```diff
@@ -0,0 +1,215 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package diagnose
+
+import (
+	"context"
+	"fmt"
+	"github.com/submariner-io/submariner-operator/internal/execute"
+	"github.com/submariner-io/submariner-operator/pkg/reporter"
+
+	"github.com/pkg/errors"
+	"github.com/submariner-io/submariner-operator/internal/cli"
+	operatorConstants "github.com/submariner-io/submariner-operator/internal/constants"
+	submv1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	"github.com/submariner-io/submariner/pkg/routeagent_driver/constants"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/kubernetes"
+)
+
+var supportedNetworkPlugins = []string{
+	constants.NetworkPluginGeneric, constants.NetworkPluginCanalFlannel, constants.NetworkPluginWeaveNet,
+	constants.NetworkPluginOpenShiftSDN, constants.NetworkPluginOVNKubernetes, constants.NetworkPluginCalico,
+}
+
+var calicoGVR = schema.GroupVersionResource{
+	Group:    "crd.projectcalico.org",
+	Version:  "v1",
+	Resource: "ippools",
+}
+
+func CheckCNIConfig(cluster *execute.Cluster) bool {
+	status := cli.NewReporter()
+
+	if cluster.Submariner == nil {
+		status.Warning(operatorConstants.SubmMissingMessage)
+
+		return true
+	}
+
+	status.Start("Checking Submariner support for the CNI network plugin")
+
+	isSupportedPlugin := false
+
+	for _, np := range supportedNetworkPlugins {
+		if cluster.Submariner.Status.NetworkPlugin == np {
+			isSupportedPlugin = true
+			break
+		}
+	}
+
+	if !isSupportedPlugin {
+		status.Failure("The detected CNI network plugin (%q) is not supported by Submariner."+
+			" Supported network plugins: %v\n", cluster.Submariner.Status.NetworkPlugin, supportedNetworkPlugins)
+		return false
+	}
+
+	status.Success("The detected CNI network plugin (%q) is supported", cluster.Submariner.Status.NetworkPlugin)
+	status.End()
+
+	return checkCalicoIPPoolsIfCalicoCNI(cluster)
+}
+
+func detectCalicoConfigMap(clientSet kubernetes.Interface) (bool, error) {
+	cmList, err := clientSet.CoreV1().ConfigMaps(metav1.NamespaceAll).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return false, errors.Wrap(err, "error listing ConfigMaps")
+	}
+
+	for i := range cmList.Items {
+		if cmList.Items[i].Name == "calico-config" {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+func checkCalicoIPPoolsIfCalicoCNI(info *execute.Cluster) bool {
+	status := cli.NewReporter()
+
+	status.Start("Trying to detect the Calico ConfigMap")
+	found, err := detectCalicoConfigMap(info.KubeClient)
+	if err != nil {
+		status.Failure("Error trying to detect the Calico ConfigMap: ", err)
```

Comment:
```suggestion
		status.Error(err, "Error detecting the Calico ConfigMap ")
```

Reply:
`status.Error` again returns an error and that error needs to be acknowledged. Using `status.failure` and `status.End` instead.